# Chapter 10 Lecture Notes

Please read chapter 10 of the textbook.

These notes take 1 - 3 lecture hours to cover.

## Dictionaries

A Python **dictionary** is a value that represents a table of data. A dictionary
is a collection of *key:value* pairs, where each key is associated with a value.
Importantly, there are *no duplicate keys*: the keys are *unique*. Because if
this restriction, looking up values by keys is extremely fast, even for large
dictionaries.

## Making Dictionaries

Suppose a university has three semesters: spring, summer, and fall. Due to the
intricacies of its database system, the spring semester is number 1, the summer
semester is numbered 2, and the fall semester is numbered 7:

| Semester | Number |
|----------|--------|
| Spring   | 1      |
| Summer   | 4      |
| Fall     | 7      |

We can represent this mapping in Python as a dictionary like this:

In [ ]:
semester_code = {'spring': 1, 'summer': 4, 'fall': 7}

print(semester_code)
print(semester_code.keys())   # print the keys
print(semester_code.values()) # print the values
print(len(semester_code))     # print the number of key:value pairs

{'spring': 1, 'summer': 4, 'fall': 7}
dict_keys(['spring', 'summer', 'fall'])
dict_values([1, 4, 7])
3


`semester_code` is a dictionary with three *key:value* pairs: `'spring': 1`,
`'summer': 4`, and `'fall': 7`. The keys are `'spring'`, `'summer'`, and
`'fall'`, and the corresponding values are `1`, `4`, and `7`.

You get the code for a semester by writing `semester_code[key]`, where `key` is
one of the key strings `'spring'`, `'summer'`, or `'fall'`:

In [ ]:
semester_code = {'spring': 1, 'summer': 4, 'fall': 7}

print(semester_code['fall'])    # 7
print(semester_code['summer'])  # 4
print(semester_code['spring'])  # 1

print(semester_code['autumn'])  # KeyError: 'autumn' is not a key

7
4
1


KeyError: 'autumn'

If look up a key that is not in the dictionary, you get a `KeyError`.

You can add new *key:value* pairs to a dictionary by assigning to them:

In [1]:
semester_code = {}            # empty dictionary
semester_code['spring'] = 1   # add some key:value pairs
semester_code['summer'] = 4
semester_code['fall'] = 7

print(semester_code)
print(semester_code.keys())   # print the keys
print(semester_code.values()) # print the values
print(len(semester_code))     # print the number of key:value pairs

{'spring': 1, 'summer': 4, 'fall': 7}
dict_keys(['spring', 'summer', 'fall'])
dict_values([1, 4, 7])
3


You can change the value associated with a key by assigning to it:

In [ ]:
semester_code = {}            # empty dictionary
semester_code['spring'] = 1
semester_code['summer'] = 4
semester_code['fall'] = 7

semester_code['fall'] = 8     # change the value of 'fall' to 8

print(semester_code)
print(semester_code.keys())   # print the keys
print(semester_code.values()) # print the values
print(len(semester_code))     # print the number of key:value pairs

{'spring': 1, 'summer': 4, 'fall': 8}
dict_keys(['spring', 'summer', 'fall'])
dict_values([1, 4, 8])
3


In general, if `d` is a dictionary:
- `d[key]` is the value associated with `key`
- `d[key] = value` sets the value associated with `key` to `value`

These operations are *very* efficient, even with large dictionaries.

## Searching Dictionaries

You can use the `in` operator to efficiently check if a key is in a dictionary:

In [ ]:
semester_code = {'spring': 1, 'summer': 4, 'fall': 7}

print('fall' in semester_code)    # True
print('autumn' in semester_code)  # False

print(1 in semester_code)         # False

True
False
False


When used with a dictionary, `in` always checks keys. `k in semester_code` is
`True` if `k` is a key in the dictionary, and `False` otherwise.

### Hashing

The `in` operator does *not* search the keys one after the other. Instead, `in`
with dictionaries use a neat trick called **hashing**. Even if a dictionary has
millions of keys, checking for a key with `in` is nearly instantaneous.

Lets sketch the basic idea. Hashing associates a random-looking **hash value**
with each key:

In [ ]:
print(hash('spring'))  # -6298311827346143009
print(hash('summer'))  # -3322306029333743668

-6298311827346143009
-3322306029333743668


Hash values may look random, but they are actually *deterministic*: the same key
always gets the same hash value.

Using some arithmetic, we can convert a hash value into a list index. Say we
have a list of length 100. Then:

In [ ]:
print(abs(hash('spring')) % 100)  # 9
print(abs(hash('summer')) % 100)  # 68

9
68


In a list of length 100, we would store `'spring'` at index 9 and `'summer'` at
index 68. To check if the list contains `'spring'`, we compute the hash value of
`'spring'` and see if `'spring`' is there.

It's possible that two different keys could hash to the same value, and so a
tie-breaking rule is needed. Many such rules have been proposed, although we
won't go into the details here.

In general, hashing is *much* faster than searching through a list of keys one
by one. This speed, combined with its relative ease of use, makes it a very
useful and practical data structure.

## Searching Dictionaries Values

Searching through the values of a dictionary is not as efficient as searching
for keys, although it is east to write:

In [ ]:
semester_code = {'spring': 1, 'summer': 4, 'fall': 7}

print(1 in semester_code.values())  # True
print(2 in semester_code.values())  # False
print(7 in semester_code.values())  # True

True
False
True


Searching values does *not* use hashing. Instead, value searches are done value
by value, as if using a loop. This is much slower than searching for keys, and
so use it with care.

## Example: A List of Words

Suppose we count how many words in [words.txt](words.txt) are the reverse of
another word. For example, "stressed" is the reverse of "desserts".

A simple, but very inefficient, way to do this is to read all the words into a
list and for each word search if it's reverse is also in the list:

In [ ]:
word_list = open('words.txt', 'r').read().split('\n')

def reverse(word):
    return ''.join(reversed(word))

def too_slow():
    count = 0
    for word in word_list:
        if reverse(word) in word_list:
            count += 1
    return count

too_slow()

886

This takes over a minute to run on my computer!

Why is it so slow? There are $n$ words in `word_list`, and for each of those $n$
words we search the entire list for its reverse. Sometimes the search for the
reverse is quick because it finds the word early in the list. But most of the
time it searches the entire list, all $n$ words, without finding the reverse.

So we search for $n$ words, and, at worst, each search checks $n$ words. In
total, that's $n \cdot n = n^2$ words we need to check. So for
[words.txt](words.txt):

In [ ]:
word_list = open('words.txt', 'r').read().split('\n')

n = len(word_list)
print(f'{n}^2 = {n**2}')

113784^2 = 12946798656


12,946,798,656 is almost 13 billion: the code checks just under 13 billion
words!

## A Dictionary of Words

We can improve the performance of the reverse-word checking program by the
storing the words in a dictionary. We'll store the words as keys, and the
correspond values will be 1s:

In [ ]:
word_dict = {}
for w in open('words.txt', 'r'):
    word_dict[w.strip()] = 1

def reverse(word):
    return ''.join(reversed(word))

def pretty_fast():
    count = 0
    for word in word_list:
        if reverse(word) in word_dict:  # word_dict instead of word_list
            count += 1
    return count

pretty_fast()

885

This is much faster than the list version: it takes less than one-tenth of a
second to run on the same computer.

This huge speedup is due to the fact that searching for a key in a dictionary
using `in` is nearly instantaneous. With a dictionary, `in` does *not* search
the words one at a time the way it does with a list.

## Example: Counting Characters in a String

Suppose you want to count characters in a string. A dictionary is a good choice
for doing this: the keys will be the characters, and the values will be the
counts.

In [ ]:
s = 'applesauce'

letters = {}
for c in s:
    if c in letters:
        letters[c] += 1
    else:
        letters[c] = 1

print(letters)

{'a': 2, 'p': 2, 'l': 1, 'e': 2, 's': 1, 'u': 1, 'c': 1}


We can put this in to a function:

In [ ]:
def count_letters(s):
    letters = {}
    for c in s:
        if c in letters:
            letters[c] += 1
        else:
            letters[c] = 1
    return letters

print(count_letters('applesauce'))

{'a': 2, 'p': 2, 'l': 1, 'e': 2, 's': 1, 'u': 1, 'c': 1}


`count_letters` works with any string, so lets count the frequency of letters in
English words:

In [ ]:
# read in words.txt as a big string
words = open('words.txt', 'r').read()
letters = {}
for c in words:
    if c not in ' \n':  # ignore spaces and newlines
        if c in letters:
            letters[c] += 1
        else:
            letters[c] = 1

print(letters)

{'a': 68574, 'h': 20186, 'e': 106752, 'd': 34548, 'i': 77392, 'n': 60505, 'g': 27832, 's': 86526, 'l': 47003, 'r': 64963, 'v': 9186, 'k': 9366, 'w': 8533, 'o': 54538, 'f': 12706, 'b': 17794, 'c': 34281, 'u': 31151, 't': 57029, 'm': 24739, 'p': 25789, 'y': 13473, 'x': 2700, 'j': 1780, 'z': 3750, 'q': 1632}


We can convert the dictionary to a list like this:

In [ ]:
# read in words.txt as a big string
words = open('words.txt', 'r').read()
letter_count = {}
for c in words:
    if c not in ' \n':  # ignore spaces and newlines
        if c in letter_count:
            letter_count[c] += 1
        else:
            letter_count[c] = 1

# read the key:value pairs into a list
letter_list = []
for c in letter_count:
    pair = [letter_count[c], c]
    letter_list.append(pair)

# sort from smallest count to biggest count
letter_list.sort()

# reverse so that the biggest count is first
letter_list.reverse()

for pair in letter_list:
    print(f'{pair[1]}: {pair[0]}')


e: 106752
s: 86526
i: 77392
a: 68574
r: 64963
n: 60505
t: 57029
o: 54538
l: 47003
d: 34548
c: 34281
u: 31151
g: 27832
p: 25789
m: 24739
h: 20186
b: 17794
y: 13473
f: 12706
k: 9366
v: 9186
w: 8533
z: 3750
x: 2700
j: 1780
q: 1632


## Questions

1. Why must dictionary keys be unique?

2. If `d` is a dictionary that does *not* have the key `k`, what happens when
   you call `d[k]`?

3. What does this print?

   ```python
   d ={'a': 3, 'b': 4}
   d['a'] = 4
   d['a'] = 5
   print(d['a'])
   ```

4. If `d` is a dictionary, what expression will give you all of the keys in `d`?
   What about all of the values?

5. Why is searching for a key in a dictionary faster than searching for a value?